packages/flutter_tools/test/general.shard/ios/ios_device_install_test.dart

```diff
@@ -0,0 +1,186 @@
+        ...kDyLdLibEntry,
+      })
+    ]);
+    final IOSDevice device = setUpIOSDevice(processManager: processManager);
+    final bool wasUninstalled = await device.uninstallApp(iosApp);
+
+    expect(wasUninstalled, true);
+    expect(processManager.hasRemainingExpectations, false);
+  });
+
+  testWithoutContext('IOSDevice.isAppInstalled catches ProcessException from ios-deploy', () async {
+    final IOSApp iosApp = PrebuiltIOSApp(projectBundleId: 'app');
+    final FakeProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
+      FakeCommand(command: const <String>[
+        'ios-deploy',
+        '--id',
+        '1234',
+        '--exists',
+        '--bundle_id',
+        'app',
+      ], environment: const <String, String>{
+        'PATH': '/usr/bin:null',
+        ...kDyLdLibEntry,
+      }, onRun: () {
+        throw const ProcessException('ios-deploy', <String>[]);
+      })
+    ]);
+    final IOSDevice device = setUpIOSDevice(processManager: processManager);
+    final bool isAppInstalled = await device.isAppInstalled(iosApp);
+
+    expect(isAppInstalled, false);
+    expect(processManager.hasRemainingExpectations, false);
+  });
+
+  testWithoutContext('IOSDevice.installApp catches ProcessException from ios-deploy', () async {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    final IOSApp iosApp = PrebuiltIOSApp(
+      projectBundleId: 'app',
+      bundleDir: fileSystem.currentDirectory,
+    );
+    final FakeProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
+      FakeCommand(command: const <String>[
+        'ios-deploy',
+        '--id',
+        '1234',
+        '--bundle',
+        '/',
+        '--no-wifi',
+      ], environment: const <String, String>{
+        'PATH': '/usr/bin:null',
+        ...kDyLdLibEntry,
+      }, onRun: () {
+        throw const ProcessException('ios-deploy', <String>[]);
+      })
+    ]);
+    final IOSDevice device = setUpIOSDevice(processManager: processManager);
+    final bool wasAppInstalled = await device.installApp(iosApp);
+
+    expect(wasAppInstalled, false);
+  });
+
+  testWithoutContext('IOSDevice.uninstallApp catches ProcessException from ios-deploy', () async {
+    final IOSApp iosApp = PrebuiltIOSApp(projectBundleId: 'app');
+    final FakeProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
+      FakeCommand(command: const <String>[
+        'ios-deploy',
+        '--id',
+        '1234',
+        '--uninstall_only',
+        '--bundle_id',
+        'app',
+      ], environment: const <String, String>{
+        'PATH': '/usr/bin:null',
+        ...kDyLdLibEntry,
+      }, onRun: () {
+        throw const ProcessException('ios-deploy', <String>[]);
+      })
+    ]);
+    final IOSDevice device = setUpIOSDevice(processManager: processManager);
+    final bool wasAppUninstalled = await device.uninstallApp(iosApp);
+
+    expect(wasAppUninstalled, false);
+  });
+}
+
+IOSDevice setUpIOSDevice({
```

Comment:
We could eventually put these helpers in a new file somewhere. I think its a lot easier to track what is going on versus the `setUp` callbacks

Reply:
I'm not a big fan of having test helpers in another file, I find it makes it less obvious (more magical) all the internals of the test objects, and it if it's shared, it becomes difficult if a new test file needs slightly different behavior.

Reply:
Yeah, thats a good point. How do you feel about the `.test` constructors? Seems like that avoids some of the mystery?

Reply:
Yeah, they're very readable. It seems unfortunate to have to put test code in the source, but maybe a necessary evil.

Reply:
Personally I prefer these in `setUp`.  Anyone familiar with dart code or general unit testing can understand `setUp` and know where to look for it, but no one except a Flutter tool developer who has read this code before knows what `setUpIOSDevice` is without clicking through to see where it is.  (Intellectually, not literally) it's like making a little utility class wrapping 2 lines of code that only Flutter tool engineers understand without clicking through and reading, instead of using standard library calls that all dart developers understand.

That being said, it's not a big deal and I'll defer to the group.  Why do _you_ find it a lot easier to track it in a method?

Reply:
My concern with setUp is that it starts out okay for small files but really doesn't scale as they inevitable get bigger. Once more helper functions get introduced, and these get nested within the group/setUp/test scope somewhere, it makes it very difficult to determine the initialization order.

I don't know if hoisting the helper functions to the top level scope is necessarily the best solution, but it does make the test easier to read

Reply:
I think if we avoid both nesting `group()`s and global (i.e. not in a group) `setUp()`s, the tests would be a lot more manageable.